blockcache/blockcache.go

```diff
@@ -0,0 +1,117 @@
+package blockcache
+
+import (
+	"sync"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/neutrino/cache"
+	"github.com/lightninglabs/neutrino/cache/lru"
+)
+
+// BlockCache is an lru cache for blocks.
+type BlockCache struct {
+	Cache     *lru.Cache
+	blockMtxs *blockMutex
+}
+
+// NewBlockCache creates a new BlockCache with the given maximum capacity.
+func NewBlockCache(capacity uint64) *BlockCache {
+	return &BlockCache{
+		Cache: lru.NewCache(capacity),
+		blockMtxs: &blockMutex{
+			mutexMap: make(map[chainhash.Hash]*blockEntry),
+		},
+	}
+}
+
+// GetBlock first checks to see if the BlockCache already contains the block
+// with the given hash. If it does then the block is fetched from the cache and
+// returned. Otherwise the getBlockImpl function is used in order to fetch the
+// new block and then it is stored in the block cache and returned.
+func (bc *BlockCache) GetBlock(hash *chainhash.Hash,
+	getBlockImpl func(hash *chainhash.Hash) (*wire.MsgBlock,
+		error)) (*wire.MsgBlock, error) {
+
+	bc.LockHash(*hash)
+	defer bc.UnlockHash(*hash)
+
+	// Create an inv vector for getting the block.
+	inv := wire.NewInvVect(wire.InvTypeWitnessBlock, hash)
+
+	var block *wire.MsgBlock
```

Comment:
Do we need this here? Could just do `block, err := getBlockImpl(hash)` on  line 56?